src/librustc/ty/instance.rs

```diff
@@ -374,6 +375,29 @@ impl<'tcx> Instance<'tcx> {
     pub fn is_vtable_shim(&self) -> bool {
         if let InstanceDef::VtableShim(..) = self.def { true } else { false }
     }
+
+    pub fn polymorphize(&mut self, tcx: TyCtxt<'tcx>) {
+        if tcx.sess.opts.debugging_opts.polymorphize {
+            debug!("polymorphize: running polymorphization analysis");
+            let results = tcx.used_generic_parameters(self.def_id());
+
+            debug!("polymorphize: results={:?}", results);
+            let polymorphized_substs = InternalSubsts::for_item(tcx, self.def_id(), |param, _| {
+                if let ty::GenericParamDefKind::Type { .. } = param.kind {
+                    if !results.contains(param.index) {
+                        // If parameter is a type parameter and is not used, then use the identity
+                        // for this parameter.
+                        return tcx.mk_param_from_def(param);
+                    }
+                }
```

Comment:
This should be uniformly done for both type and `const` generics.

Reply:
I've fixed this in 4dc7f3f.